### ISSUE TITLE
Bug: [@typescript-eslint/strict-boolean-expressions] False negative for "return ... && ..."

### ISSUE THREAD

**Issue by ilyub** [typescript-eslint/typescript-eslint#5159]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play#ts=4.7.2&sourceType=module&code=LAKAZgrgdgxgLgSwPZQARgBQA8BcqBGSSANgKYCGUANKgJ4D8eAznAE4JQDmAlKgN6hUqBGFTZefVAF9Qg4aIy0J02SCEixWVADJtdZTJBzWpOBFZotuugG5Qh0JFiIUqTtjyESFanUaoWdi4JOQ1xfhUjNXkxJQiHaLCrPTjJBKETMwtUZNQAQlo7ECkgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZfKsugIwHs+SAIZN6AD2LQUySnyaoMkAO5DoTSODABfEFqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
function f(x: boolean, y?: string) {
  if (x) { }

  if (y) { }

  if (x && y) { }

  return x && y;
}

function g(x: boolean, y?: string) {
  if (x) { }

  if (y) { }

  if (x && y) { }

  return x && !y;
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/strict-boolean-expressions": ["error", ...<options>],
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

It should report "y" in this line:
```ts
return x && y;
```

The reason:

Note that "y" is assumed to be a part of a boolean expression and checked when it is used with boolean "!" expression:
```ts
return x && !y;
```

I think that using "y" with boolean "&&" or boolean "||" operators is also enough to assume that it is a part of boolean expression.

### Actual Result

No warning in this line:
```ts
return x && y;
```

### Versions

| package | version |
| -- | -- |
| `@typescript-eslint/eslint-plugin` | `5.27.1` |
| `@typescript-eslint/parser` | `5.27.1` |
| `TypeScript` | `4.7.2` |
| `ESLint` | `8.15.0` |
| `node` | `web` |

**Comment by bradzacher** [comment]
This is actually entirely on purpose.

For the rule we consider the context of the boolean expression.

In the context of a "boolean location" like an `if`, we strictly check every side of the boolean expression because we know that (a) the return type is "supposed" to be a boolean, and (b) TS only cares if it's "truth" /"falsy".

In the context of another expression we intentionally only check the "boolean" sides of the expression - however the "returned" sides of the expression we ignore because we assume that the type of the expression correctly covers the thing.

I.e.
```ts
const c: boolean = left && right;
//                  ^^^^ covered by the rule
//                          ^^^^^ covered by TS
```
